File: holoken-core/src/main/java/com/holokenmod/calculation/GridPreviewCalculationService.java
package com.holokenmod.calculation;

import com.holokenmod.creation.GridCalculator;
import com.holokenmod.grid.Grid;
import com.holokenmod.options.GameVariant;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Future;
import java.util.concurrent.FutureTask;
import java.util.function.Function;

public class GridPreviewCalculationService {
	
	private final Map<GameVariant, Grid> grids = new HashMap<>();
	
	public GridPreviewCalculationService() {
	
	}
	
	public Future<Grid> getOrCreateGrid(GameVariant variant) {
		FutureTask<Grid> future = new FutureTask<>(() -> grids.computeIfAbsent(variant, computeVariant()));
		
		new Thread(future).start();
		
		return future;
	}
	
	private Function<GameVariant, Grid> computeVariant() {
		return (variant) -> {
			final GridCalculator creator = new GridCalculator(variant.getGridSize());
			
			return creator.calculate();
		};
	}
	
	public Grid getGrid(GameVariant gameVariant) {
		return grids.get(gameVariant);
	}
}

File: holoken-core/src/main/kotlin/com/holokenmod/calculation/GridPreviewCalculationService.kt
package com.holokenmod.calculation

import com.holokenmod.creation.GridCalculator
import com.holokenmod.grid.Grid
import com.holokenmod.options.GameVariant
import java.util.WeakHashMap

class GridPreviewCalculationService {
    private val grids: MutableMap<GameVariant, Grid> = WeakHashMap()

    suspend fun getOrCreateGrid(variant: GameVariant): Grid {
        grids[variant]?.let {
            return it
        }

        val grid = GridCalculator(variant).calculate()
        grids[variant] = grid

        return grid
    }

    fun getGrid(gameVariant: GameVariant): Grid? {
        return grids[gameVariant]
    }
}
